feat: add jump to category dropdown in tools dashboard

janvi01 opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow our contribution guidelines
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**

Added a Category menu dropdown in the tool's dashboard to enable users to directly jump to the required Category section they want. Below is the screenshot that depicts the implementation.

<img width="1792" alt="Screenshot 2023-02-13 at 11 11 50 AM" src="https://user-images.githubusercontent.com/34344234/218387533-a4c7ef56-1c1f-4771-a887-ce05348e1981.png">

Does this look more appropriate (spacing between category and arrow)?

<img width="1792" alt="Screenshot 2023-02-13 at 11 03 42 AM" src="https://user-images.githubusercontent.com/34344234/218389080-c7f9d24a-9491-4fcb-af8e-d8fe2f06cb7e.png">


For navigation, I have added the category name as `div id` as of now since I couldn't find any appropriate property in `tools.json`. We have `tags` in `categorylist.js` but that is not used for mapping categories in `ToolsList.js`. Please let me know of any viable solution for this.


**Related issue(s)**
Fixes #1319

janvi01 committed:
add: jump to category dropdown in tools dashboard

akshatnema reviewed (changes requested):
Hey @janvi01, thanks for working on this issue. Here are certain points you need to resolve in their PR:

![image](https://user-images.githubusercontent.com/76521428/218763951-958282f9-3227-48f0-896b-3458527886f8.png)

- Please change the name of the dropdown from `Category` to `Jump to Category`.
- Make the dropdown scrollable in itself with limited height only. It shouldn't be this long enough.
- The dropdown should get closed when clicked outside.
- The mobile view of the Filter Bar should be corrected properly, It should be responsive.   
![image](https://user-images.githubusercontent.com/76521428/218765850-b519c926-2f13-4b0c-941a-7e53c07a8b60.png)

janvi01 committed:
fix: certain require changes like dropdown height, responsive etc.

janvi01 commented:
Screenshot - 

<img width="1792" alt="Screenshot 2023-02-15 at 12 45 03 PM" src="https://user-images.githubusercontent.com/34344234/218959038-3e6b72d5-60af-4d55-b16a-18045064b2ea.png">

akshatnema commented on components/tools/CategoryDropdown.js:
```diff
@@ -0,0 +1,16 @@
+      {categoryList.map((data, index) => {
+        return (
+        <div onClick={() => setopenCategory(false)}>
+          <a href={`#${data.name}`} key={index} class="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white" role="menuitem" tabindex="-1" id="menu-item-0">{data.name}</a>
```
```suggestion
          <a href={`#${data.name}`} key={index} class="block px-4 py-2 hover:bg-gray-100">{data.name}</a>
```

we don't have a dark mode on our website, so don't add any dark mode CSS properties.

akshatnema commented on components/tools/ToolDashboard.js:
```diff
@@ -109,6 +125,21 @@ export default function ToolDashboard() {
                     </div>
                 </div>
+                <div className='flex w-1/3 lg:w-1/5 gap-5 h-auto'>
+                    <div className="relative w-full h-auto" ref={categoryRef}>
```
I think the `ref` should be applied to the Dropdown menu only, not on the whole section of `Jump to Category`.

akshatnema reviewed (changes requested):
![image](https://user-images.githubusercontent.com/76521428/218982996-d076a962-1e16-4e4b-b44e-edd7a169961a.png)

The gap between the dropdown menu and the button is somewhat large. Make it same as of Filter button and it's onClick Modal.

![image](https://user-images.githubusercontent.com/76521428/218985322-0e33080f-b6a5-468c-a1eb-7eee34f9a846.png)

As I mentioned earlier, the mobile view is not responsive yet. For mobile view, you have to align the Filter and Jump To Category buttons on the same line and the Search bar on next line. Hope you have understood, what I meant...

Also, try to customize the scroll bar of the dropdown to make it a little bit narrow, using Tailwind only, because right now, it looks slightly off in the looks.

janvi01 committed:
Merge branch 'master' into ToolsJumpMenu

janvi01 committed:
fix: certain require changes

janvi01 commented:
Fixed the gap between the dropdown menu and the button, making it responsive for mobile.

To make the dropdown compact, I have removed the scrollbar. Does that look good? `scrollbar-thin` is not working as I think we need a separate plugin for that.

screenshot -

<img width="1792" alt="Screenshot 2023-02-16 at 1 18 57 PM" src="https://user-images.githubusercontent.com/34344234/219302433-ffe656ab-e425-4d32-a0ee-92b5161ba728.png">

akshatnema reviewed (commented):
![image](https://user-images.githubusercontent.com/76521428/220878950-7b2fa016-833a-4ddb-a9e0-1c1f8f6163c2.png)

Make the height of the Filter button and Category button the same as that of the Search Bar in the mobile view.

And, please don't remove the scroll from the menu as it will be needed to show the users that we have a list of options in the menu. We can customize it later for the whole website if needed.

janvi01 committed:
update require changes

janvi01 commented:
screenshot -

<img width="1792" alt="Screenshot 2023-02-24 at 4 59 11 PM" src="https://user-images.githubusercontent.com/34344234/221168479-a2316ad9-e1df-4b79-97e0-f2daa5313e95.png">

akshatnema reviewed (changes requested):
![Screenshot_2023-02-26-11-14-06-28.jpg](https://user-images.githubusercontent.com/76521428/221394517-4fc569a0-cf6c-4ee8-9cb6-46104fce3af9.jpg)


The category drop-down when opened makes the webpage less responsive and adds an unnecessary horizontal scroll in the website. This should be removed and the drop-down shouldn't extend beyond the screen display in mobile view.

Also, I see that in mobile view, clicking back to `Jump to Category` button doesn't closes the drop-down. Make a onclick functionality for that too.

Lastly, I also see that all the categories are shown in the drop-down whether that category of tools are present in the website or not. That's quite wrong as we can't point to those categories that doesn't exist in the webpage. So, do make a check that include those categories that are present in the webpage. (In short, include those categories for which atleast 1 tool is present in the list)

janvi01 committed:
update require changes

janvi01 commented:
I have made the dropdown responsive for mobile view and fixed the dropdown not closing back error.

Screenshot -

<img width="457" alt="Screenshot 2023-02-28 at 5 11 20 PM" src="https://user-images.githubusercontent.com/34344234/221845473-c9084fd6-48ec-487c-af6e-c38e96e323ba.png">

For the last point you mentioned, I have mapped the dropdown list with the `categorylist.js` data that is present there, how would I filter if we have tools for a particular category or not?

akshatnema commented:
> For the last point you mentioned, I have mapped the dropdown list with the categorylist.js data that is present there, how would I filter if we have tools for a particular category or not?

You can use `tools.json` data to filter out the necessary categories from the tools.

Also, please resolve the merge conflicts of the PR.

janvi01 committed:
fix: conflicts

janvi01 committed:
fix: used tools.json to list tools categories

janvi01 commented:
Hey, @akshatnema I have made the required changes, do let me know if we need any more changes.

janvi01 committed:
Merge branch 'master' into ToolsJumpMenu

akshatnema commented on components/tools/CategoryDropdown.js:
```diff
@@ -0,0 +1,17 @@
+export default function CategoryDropdown({setopenCategory}) {
+
+  return (
+  <div class="absolute z-10 w-52 lg:w-56 origin-top-right h-60 overflow-y-auto rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none" role="menu" aria-orientation="vertical" aria-labelledby="menu-button" tabindex="-1">
```
```suggestion
  <div class="absolute z-10 -left-[2rem] lg:left-0 w-52 lg:w-56 origin-top-right h-60 overflow-y-auto rounded-md bg-white shadow-lg ring-1 ring-black ring-opacity-5 focus:outline-none" role="menu" aria-orientation="vertical" aria-labelledby="menu-button" tabindex="-1">
```

To remove the scroll from smaller screens.

janvi01 replied:
I have removed that, we don't get a horizontal scrollbar anymore.

Applying `-left-[2rem]` for a smaller screen looks a bit odd, should I do that?

screenshot -

<img width="753" alt="Screenshot 2023-03-06 at 3 47 19 PM" src="https://user-images.githubusercontent.com/34344234/223082469-4fb34c51-4778-463b-8569-48c95d5caf72.png">

akshatnema replied:
Yeah, I found this solution to be not working correctly on different screens. I've another way of doing it and it is in the suggestions below.

akshatnema reviewed (changes requested):
![image](https://user-images.githubusercontent.com/76521428/222963096-c6e84247-c2cc-47f0-a49c-c8cbae78da48.png)

Also, I see a small gap in the alignment of the button and search bar in the mobile view. Can you please fix that?

janvi01 committed:
fix: buttons alignment

akshatnema commented:
![image](https://user-images.githubusercontent.com/76521428/223422786-a3299bab-601c-4723-893e-998afaa0d522.png)

@janvi01 with your latest changes, I see that the buttons are not aligned inline now in mobile view. Kindly correct it.

akshatnema commented on components/tools/ToolDashboard.js:
```diff
@@ -123,7 +139,22 @@ export default function ToolDashboard() {
+                            <ArrowDown className={`my-auto ${openCategory ? 'rotate-180' : ''}`} />
+                        </div>
+                        {openCategory && (
+                            <div className="z-20 absolute top-16">
```
```suggestion
                            <div className="z-20 absolute top-16 right-52 lg:right-0">
```

I think this will correct the scroll error.

janvi01 replied:
`lg:right-0` is making the dropdown appear below the search bar for large screens. I think `right-52` is good for all screen sizes.

janvi01 committed:
update require changes

akshatnema committed:
class corrected

akshatnema committed:
Merge branch 'master' into ToolsJumpMenu

akshatnema commented:
/rtm
